### PR TITLE
implemented API with timeout and polling interval

### PR DESCRIPTION
**Pull request by akamarouski**
Implemented "Implement API call fluentWait analogue with timeout and polling interval" #330 ticket.

Implemented a function that allows the user to expect the correct answer based on the lambda expression passed to them as a parameter. Created an WaitingRequestProps annotation through which the request interval, waiting time and type of time are indicated. Created unit-test for WaitingRequestProps annotation.

### CODE REVIEW

**Comment by SergeiZagriychuk** [comment]
let's also document this feature in carina-api documentation

File: carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java [hunk 1]
```diff
@@ -127,9 +143,61 @@ public Response callAPI() {
...
+    }
+
+    /**
+     * Waits the correct response
+     *
+     * @param isValid - passed function to check the correctness of the response
+     */
+    public Response callAPI(Predicate<Optional<Response>> isValid) {
+        WaitingRequestProps waitingRequestProps = this.getClass().getAnnotation(WaitingRequestProps.class);
+        if (waitingRequestProps == null) {
```
Comment (SergeiZagriychuk) [inline carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java:166]: check params for not-null instead

File: carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java [hunk 2]
```diff
@@ -127,9 +143,61 @@ public Response callAPI() {
...
+            throw new AnnotationMissingException("To use this method please declare @WaitingRequestProps for your AbstractApiMethod class");
+        }
+
+        ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor();
+
+        Runnable callAPITask = () -> {
```
Comment (SergeiZagriychuk) [inline carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java:172]: improve algorithm

File: carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java [hunk 3]
```diff
@@ -127,9 +143,61 @@ public Response callAPI() {
...
+    public TimeUnit getPeriodDelayUnit() {
+        return periodDelayUnit;
+    }
+
+    /**
+     * Waits the correct response
+     *
+     * @param isValid - passed function to check the correctness of the response
+     */
+    public Response callAPI(Predicate<Optional<Response>> isValid) {
```
Comment (SergeiZagriychuk) [inline carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java:164]: callAPIWithRetry()
isValid -> successCondition

File: carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java [hunk 4]
```diff
@@ -127,9 +143,61 @@ public Response callAPI() {
...
+    public int getRqDelay() {
+        return rqDelay;
+    }
+
+    public TimeUnit getPeriodDelayUnit() {
+        return periodDelayUnit;
+    }
+
+    /**
+     * Waits the correct response
```
Comment (SergeiZagriychuk) [inline carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java:160]: extend comment

**Review by SergeiZagriychuk — COMMENTED** [review]
let's also document this feature in carina-api documentation

File: carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java (hunk 2 above)
Comment (akamarouski) [reply carina-api/src/main/java/com/qaprosoft/carina/core/foundation/api/AbstractApiMethodV2.java:149]: Done

**Pull request closed without merging**